src/models/ListRule.ts

```diff
@@ -15,33 +15,140 @@ limitations under the License.
-export function recommendationToStable(recommendation: string, unstable = false): string|null {
-    if (RECOMMENDATION_BAN_TYPES.includes(recommendation)) return unstable ? RECOMMENDATION_BAN_TYPES[RECOMMENDATION_BAN_TYPES.length - 1] : RECOMMENDATION_BAN;
-    return null;
+    /// The rule specifies an "opinion", as a number in [-100, +100],
+    /// where +100 represents a user who is absolutely trusted and
+    /// -100 represents a user who is absolutely untrusted.
+    Opinion = "org.matrix.msc3845.opinion"
 }
 
-export class ListRule {
+/**
+ * All types for `m.ban`
+ */
+const RECOMMENDATION_BAN_VARIANTS = [
+    // Stable
+    Recommendation.Ban,
+    // Unstable prefix, for compatibility.
+    "org.matrix.mjolnir.ban"
+];
 
-    private glob: MatrixGlob;
+/**
+ * All types for `m.ban`
+ */
+const RECOMMENDATION_OPINION_VARIANTS: string[] = [
+    // Unstable
+    Recommendation.Opinion
+];
 
-    constructor(public readonly entity: string, private action: string, public readonly reason: string, public readonly kind: string) {
-        this.glob = new MatrixGlob(entity);
-    }
+export const OPINION_MIN = -100;
+export const OPINION_MAX = +100;
 
+// FIXME: This function is only ever called with a constant?
+export function recommendationToStable(recommendation: string): Recommendation | null {
+    if (RECOMMENDATION_BAN_VARIANTS.includes(recommendation)) return Recommendation.Ban;
+    if (RECOMMENDATION_OPINION_VARIANTS.includes(recommendation)) return Recommendation.Opinion;
+    return null;
+}
+
+/**
+ * Representation of a rule within a Policy List.
+ */
+export abstract class ListRule {
     /**
-     * The recommendation for this rule, or `null` if there is no recommendation or the recommendation is invalid.
-     * Recommendations are normalised to their stable types.
+     * A glob for `entity`.
      */
-    public get recommendation(): string|null {
-        if (RECOMMENDATION_BAN_TYPES.includes(this.action)) return RECOMMENDATION_BAN;
-        return null;
+    private glob: MatrixGlob;
+    constructor(
+        /**
+         * The entity covered by this rule, e.g. a glob user ID, a room ID, a server domain.
+         */
+        public readonly entity: string,
+        /**
+         * A human-readable reason for this rule, for audit purposes.
+         */
+        public readonly reason: string,
+        /**
+         * The type of entity for this rule, e.g. user, server domain, etc.
+         */
+        public readonly kind: EntityType,
+        /**
+         * The recommendation for this rule, e.g. "ban" or "opinion".
+         */
+        public readonly recommendation: Recommendation | null) {
+        this.glob = new MatrixGlob(entity);
     }
 
+    /**
+     * Determine whether this rule should apply to a given entity.
+     */
     public isMatch(entity: string): boolean {
         return this.glob.test(entity);
     }
 }
+
+/**
+ * A rule representing a "ban".
+ */
+export class ListRuleBan extends ListRule {
```

Comment:
Should the constructors of ListRule, ListRuleOpinion and ListRuleBan be exposed? Surely you want them to go through some common constructor (well it's a factory method isn't it) that will give you one or the other?

Reply:
That's the idea, yes. I'm planning to create this factory (and possibly hide the constructors) once I have a better idea of how that code will be used.

Reply:
Factory created.